### ISSUE TITLE
Docker events command do not work after 15 minutes 

### ISSUE THREAD

**Issue by redlockio** [docker/docker#15874]
Description of problem:

I'm using "docker events" command to monitor docker events in real time. After about 15 minutes, I don't see any docker events even though I am running commands like docker run etc. Note that I am using docker over TCP port. 
I wrote a Java code which uses Apache HTTP Client found that even apache client doesn't see any data from the docker TCP server. I then ran Wireshark and can confirm that there are no packets are getting sent by docker TCP server for the HTTP connection docker event command creates. 

`docker version`:

Client version: 1.7.1
Client API version: 1.19
Go version (client): go1.4.2
Git commit (client): 786b29d
OS/Arch (client): windows/amd64
Server version: 1.8.1
Server API version: 1.20
Go version (server): go1.4.2
Git commit (server): d12ea79
OS/Arch (server): linux/amd64

`docker info`:

Containers: 163
Images: 128
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 454
 Dirperm1 Supported: true
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 3.16.0-4-amd64
Operating System: Debian GNU/Linux 8 (jessie)
CPUs: 1
Total Memory: 1000 MiB
Name: XXX
ID: XXX
WARNING: No memory limit support
WARNING: No swap limit support
`uname -a`:

1 SMP Debian 3.16.7-ckt11-1 (2015-05-24) x86_64 GNU/Linux
Environment details (AWS, VirtualBox, physical, etc.):
Server- Digital Ocean

How reproducible:

Always

Steps to Reproduce:
1. Run docker events command
2. Perform some docker events e.g. run, delete etc 
3. Wait for 15 minutes

Actual Results:

docker event command no longer shows new events after 15 minutes 

Expected Results:

Docker events command should always show docker events 
Additional info:

**Comment by LK4D4** [comment]
@redlockio Could you try to connect another client to docker events when this occured? Also could you try unix socket? Thanks!

**Comment by redlockio** [comment]
@LK4D4  yes, I've two clients connected- one using apache HTTP client and another client is standard docker client. I will try with unix sockets too. thanks

**Comment by LK4D4** [comment]
I'll try to reproduce tomorrow too.

**Comment by redlockio** [comment]
Thanks. I just verified that if events keep flowing in, docker events command work just fine. I wrote following script to generate events every 1 second- 

#!/bin/bash
while :
do
        echo "Press [CTRL+C] to stop.."
        ./docker -H tcp://ip:port run some-image
        sleep 1
done

docker client running on windows was able to show events for more than three hours. So I stopped the script and didn't generate any event for 20 minutes. When I started the script again, docker events command didn't work. 

So to summarize, the issue occurs when no event has been generated for 15-20 minutes.

**Comment by LK4D4** [comment]
@redlockio I'm trying right now.

**Comment by LK4D4** [comment]
It worked for me. I wonder if it's something with tcp timeouts. So servers side socket is just silently closed.

**Comment by redlockio** [comment]
thanks for trying out, I am monitoring state of the socket via netstat command right now and will update this thread with my findings.

**Comment by redlockio** [comment]
I can confirm that TCP socket is still in established state after 20 minutes. I am monitoring data through wireshark but don't see any packets coming in from the socket attached to the tcp connection created by docker events command. 

This could perhaps be an environment issue. I am running Docker daemon on digital ocean. Let me try and see what happens with EC2 and Google compute.

**Comment by bprodoehl** [comment]
I've seen this sort of thing with other long-lived sockets when going through a masquerading NAT.  You mention Windows on the client side, and DigitalOcean.  What's the layout?  Have you tried running docker events on the DigitalOcean side?  This SO comment describes pretty well what I've seen: http://stackoverflow.com/questions/1480236/does-a-tcp-socket-connection-have-a-keep-alive/1480267#1480267

**Comment by redlockio** [comment]
@bprodoehl you may be onto something thing. I am running docker client on Windows at my home machine and the server is far away in digitial ocean. Let me see what happens when everything is within a same network so that there are no routers involved.

**Comment by thaJeztah** [comment]
@redlockio can you keep us posted when you had the time to test this locally? Thanks!

**Comment by cpuguy83** [comment]
Any news?

**Comment by redlockio** [comment]
Sorry about the delay in responding. I can confirm that when there is no
router/gateway etc in between docker daemon and client listening on events,
everything works fine.

**Comment by thaJeztah** [comment]
Thanks, @redlockio! I'll close this issue then, because I think this is not a docker issue, but feel free to comment if you think I closed by mistake
